OCamlformat RPC

voodoos opened the pull request:
This PR add support for formatting of types in popups. Types provided by Merlin are not always nicely formatted and module signatures often wrap because it exceeds the maximum number of columns of a VSCode popup (64 characters on my setup). The idea is to use OCamlformat to format these types, but starting a new process would be to heavy for that usage.

In a joint effort with the OCamlformat team this PR acheive type formatting by spawning an "OCamlformat RPC server" and then communicating with it using a simple protocol. The protocol is vendored for now (in the file `ocamlformat_rpc_lib.ml` but will be made available as a separate and common package.

In case of failure to start the server, negotiate a version or receive a response we log some information and fallback to the non-formatted type.

## Todo
- [x] Wait for OCaml-RPC lib to be published as a separate package and remove the vendored copy.
- [x] Change the program name to the definitive name of the OCamllsp-RPC server when known. (for now it is a hard path to my copy of OCamlformat).

## Related works
- This [OCamlformat PR](https://github.com/ocaml-ppx/ocamlformat/pull/1586) tracks the progress on the RPC library and the server implementation.
- This [mdx PR](https://github.com/realworldocaml/mdx/pull/316) tracks the progress on implementing a client similar to the one of this PR to format code blocks in mdx.

rgrinberg commented on ocaml-lsp-server/src/ocamlformat_rpc_lib.ml:
```diff
@@ -0,0 +1,220 @@
+(*      This source code is licensed under the MIT license found in       *)
+(*      the LICENSE file in the root directory of this source tree.       *)
+(*                                                                        *)
+(**************************************************************************)
```
If this is vendored code, it should live in the vendored directory.

voodoos replied:
It should disappear before we merge the PR. But I will move it to the vendor dir if it doesn't at the end.

rgrinberg replied:
Sounds good.

rgrinberg commented on ocaml-lsp-server/src/ocamlformat_rpc.ml:
```diff
@@ -0,0 +1,69 @@
+  | Running of process
+  | Errored
+
+let running_process : state ref = ref Uninitialized
```
This state should exist inside the LSP server instance.

voodoos replied:
You mean that it should be a part of `State.t` ?

rgrinberg replied:
Yes.

rgrinberg commented on ocaml-lsp-server/src/ocamlformat_rpc.ml:
```diff
@@ -0,0 +1,69 @@
+  let pid = Unix.process_pid (input, output) in
+  let open Result.O in
+  Ocamlformat_rpc_lib.pick_client ~pid input output supported_versions
+  >>| (fun client ->
```
`let+`

rgrinberg commented on ocaml-lsp-server/src/ocamlformat_rpc_lib.ml:
```diff
@@ -0,0 +1,220 @@
+      | `Halt ->
+        Error
+          (`Msg
+            "OCamlformat-RPC did not respond. Check that a compatible version \
```
We should probably mentioned what minimum version is required.

voodoos replied:
cc @gpetiot

gpetiot replied:
done

gpetiot commented:
`ocamlformat-rpc-lib` is now available on opam!

rgrinberg commented on ocaml-lsp-server/src/ocaml_lsp_server.ml:
```diff
@@ -336,7 +336,20 @@ let hover (state : State.t) { HoverParams.textDocument = { uri }; position; _ }
+      | Error (`Msg message) ->
+        (* We log OCamlformat errors and display the unformated type *)
+        Log.log ~section:"ocamlformat" (fun () ->
+            Log.msg "An error occured while querying ocamlformat"
```
Note that this log is mostly used for debugging. Users will not see these messages unless they open the debug pane. If you want users to see this error, use the `ShowMessage` notification.

rgrinberg commented on ocaml-lsp-server/src/ocamlformat_rpc.ml:
```diff
@@ -0,0 +1,172 @@
+            client
+        with
+        | Error (`Msg msg) ->
+          Log.log ~section:"ocamlformat" (fun () ->
```
I'd move this back to the main thread. it's better to write log messages from a single thread if possible.

rgrinberg commented on ocaml-lsp-server/src/ocamlformat_rpc.ml:
```diff
@@ -0,0 +1,172 @@
+  | Running ({ pid; io_thread; _ } as p) -> (
+    (* Is the process still running ? *)
+    match
+      Scheduler.async io_thread (fun () -> Unix.waitpid [ WNOHANG ] pid |> fst)
```
You should use `Scheduler.wait_for_process` here. We do the waiting for all processes in a special thread.

voodoos replied:
I tried, but it raise the following exception if I use it twice:
```
[Trace - 10:37:25 AM] Received response 'textDocument/hover - (10)' in 2ms. Request failed: uncaught exception (-32603).
Error data: {
    "exn": "File \"fiber-unix/src/scheduler.ml\", line 418, characters 28-34: Assertion failed",
    "backtrace": 
"Raised at Fiber_unix__Scheduler.Process_watcher.Process_table.add in file \"fiber-unix/src/scheduler.ml\", line 418, characters 28-40
Called from Fiber_unix__Scheduler.Process_watcher.register in file \"fiber-unix/src/scheduler.ml\", line 441, characters 4-31
Called from Fiber_unix__Scheduler.wait_for_process in file \"fiber-unix/src/scheduler.ml\", line 558, characters 2-71
Called from Fiber.Execution_context.run_jobs in file \"submodules/dune/src/fiber/fiber.ml\", line 204, characters 8-13\n"
}```

It looks like `Swheduler.waitpid` always tries to add the process to the watchlist `Table` and that fails when it is already in the table.

voodoos replied:
Hum I think I see it: it also uses `Unix.waitpid` which is blocking. So in fact I should not try to use it like the non-blocking `Unix.waitpid [ WNOHANG ]`.

I should not call `Scheduler.wait_for_process` each time I want to query ocf to check that it is still running.
I should use `Scheduler.wait_for_process` as a watcher that will do something if ocf stops running.

Is that right ?

rgrinberg replied:
dune.ml should give you an example of how it should work. Basically, when your RPC is in the running state, it should be roughly this:

```ocaml
Fiber.fork_and_join_unit
  (fun () -> (* code to interact with ocamlformat rpc *))
  (fun () -> Scheduler.wait_for_process pid)
```

Additionally, there should be a `Ocamlformat_rpc.run` function that will return a fiber that should complete when RPC is completely finished and everything is cleaned up.

Let me know if that's not clear and we can walk through this in a call.

rgrinberg commented on ocaml-lsp-server/src/ocamlformat_rpc.mli:
```diff
@@ -0,0 +1,7 @@
+
+val stop : t -> unit Fiber.t
+
+val query_type : t -> string -> (string, [> `Msg of string ]) result Fiber.t
```
`string` should be labelled.

voodoos replied:
I used a type alias to also make the return type clearer:

`val query_type : t -> type_ -> (type_, [> `Msg of string ]) result Fiber.t`

rgrinberg commented on ocaml-lsp-server/src/ocamlformat_rpc.ml:
```diff
@@ -0,0 +1,172 @@
+open Import
+
+type process =
+  { pid : int
```
I'd prefer `Stdune.Pid.t` here.

rgrinberg commented on ocaml-lsp-server/src/ocamlformat_rpc.ml:
```diff
@@ -0,0 +1,172 @@
+          Log.log ~section:"ocamlformat" (fun () ->
+              Log.msg "An error occured while configuring ocamlformat"
+                [ ("msg", `String msg) ])
+        | Ok () -> ())
```
This short clause should be first.

rgrinberg commented on ocaml-lsp-server/src/ocamlformat_rpc.ml:
```diff
@@ -0,0 +1,172 @@
+
+let start ?io_thread t =
+  let open Fiber.O in
+  try
```
This `try` is suspicious. It will not catch exceptions raised by fibers. You're probably looking for `Fiber.map_reduce_errors`.

rgrinberg commented on ocaml-lsp-server/src/ocamlformat_rpc.ml:
```diff
@@ -0,0 +1,172 @@
+    let output = Unix.out_channel_of_descr stdin in
+    let* io_thread =
+      (* Don't create a new thread if there is already one. *)
+      (* This can happen if the Ocamlformat server crashed. *)
```
I'd prefer not to reuse the threads between processes unless there's a concrete performance improvement that is totally necessary. The issue is that we might be stuck arbitrarily long waiting for existing tasks to complete. If we create a new thread, we'll always get new responses promptly.

rgrinberg commented on ocaml-lsp-server/src/ocamlformat_rpc.ml:
```diff
@@ -0,0 +1,172 @@
+    let argv = [| bin |] in
+    let pid, stdout, stdin =
+      let stdin_i, stdin_o = Unix.pipe () in
+      let stdout_i, stdout_o = Unix.pipe () in
```
I don't see these fd's being closed anywhere. That seems wrong.

rgrinberg commented on ocaml-lsp-server/src/ocamlformat_rpc.ml:
```diff
@@ -0,0 +1,172 @@
+      Scheduler.async process.io_thread (fun () ->
+          Ocamlformat_rpc_lib.format typ process.client)
+    with
+    | Error `Stopped -> Fiber.return @@ Error (`Msg "Thread stopped")
```
Is this error message going to be visible to the user? If so,

* Under what circumstances should the user expect this?
* The error message should be more descriptive as to what is happening.

Same for the `"No process"` message.

voodoos commented:
About error messages, I think most of them will stay as debug messages.
The only user-facing message should be an information  diagnostic showing when OCaml-LSP cannot find the `ocamlformat-rpc` binary and the user should install it for full functionality.

OCamlformat crashes and errors should be kept in the debug log: there are useful for devs to debug issues but useless to the users and unrelated to their programs.

voodoos commented:
I used the same model as for the Dune RPC.
I start the OCamllsp_rpc server and wait for it to finish in `Ocaml_lsp_server.start`:

```ocaml
Fiber.parallel_iter
          ~f:(fun f -> f ())
          [ (fun () -> Server.start server)
          ; (fun () ->
              let* _state = Ocamlformat_rpc.run ocamlformat_rpc in
              (* TODO ulysse Handle errors*)
              Fiber.return ())
          ; (fun () -> ...  Dune.run dune  ... )
          ]
```

However I have a problem to stop it: you appear to stop the DRPC client after this first fiber returns:
```ocaml
      match !dune with
      | None -> Fiber.return ()
      | Some dune -> Dune.stop dune)
```

But I don't understand why the first fiber would return before the process is stopped because it is waiting for the process to stop to complete ? In any case I had that problem with OCamlformat server and decided to stop it explicitely when the Server does:

```ocaml
Fiber.parallel_iter
          ~f:(fun f -> f ())
          [ (fun () ->
              let* () = Server.start server in
              (* When server exits we also stop ocamlformat rpc *)
              Ocamlformat_rpc.stop ocamlformat_rpc)
          ; (fun () ->
              let* _state = Ocamlformat_rpc.run ocamlformat_rpc in
              (* TODO ulysse Handle errors*)
              Fiber.return ())
          ; (fun () -> ...  Dune.run dune  ... )
          ]
```

This way the second item in the parallel iter does complete because it watches the process exit and the process is exited by the stopping of the server in the first item of the parallel iter.

Is that the right thing to do or did I miss something essential ?

(Also when I use Dune master the ocamllsp process does not exits when asking a reload of the server in vscode, process are multiplying , which does not happen with a version of Dune without the rpc server)

rgrinberg commented:
> Is that the right thing to do or did I miss something essential ?

You aren't missing anything.  I already had this fixed in a branch that I merged yesterday.

voodoos commented:
> > Is that the right thing to do or did I miss something essential ?
> 
> You aren't missing anything. I already had this fixed in a branch that I merged yesterday.

Great, I will have one last pass and it will be ready for a (hopefully) final review.

voodoos commented:
I refactored the module a bit.

I also added a user notification (`ShowNotification`) showing when the `ocamlformat-rpc` binary is not found.
However I do not show other errors, they are logged only if debug is enabled.

Some types will fail to be formatted by ocamlformat: Merlin prints types which may be disambiguated with a stamp like in `t/1 -> t/2` and this is not valid syntax.
In this case a non-formatted type will be showed to the user as before.

@gpetiot is working on supporting this syntax in ocamlformat.

rgrinberg commented:
Looks much better now! There are still some opaque error messages that need fixing (or removing)

voodoos commented:
I reworked the login.

Also, is it expected that tests fail ? They do not happen on master, but I don't understand how the failures could be related to my changes.

rgrinberg commented:
> running the tests spawns a huge number of hard-to-kill ocamlformat-rpc instances, I will have to investigate that. Tests also fail on CI because ocamlformat-rpc is not installed.

That's definitely problematic. I forgot to mention that we should delay launching the process until we actually request it to format something.

voodoos commented:
I decided to use two `Ivars` in ad5e84792384e2c12bedc722ba4a0a0a37e976a9 to only start of the process when the first query is done and then perform that query as soon as the process started.

rgrinberg commented:
Looks good. I modified the code to us two ivars instead of the mvar. I'm more confident with an ivar that we won't double write when we shouldn't.
